docs/man/man1/sendkeys.1
'\" t
.\"     Title: sendkeys
.\"    Author: [see the "AUTHOR" section]
.\" Generator: DocBook XSL Stylesheets v1.79.1 <http://docbook.sf.net/>
.\"      Date: 04/30/2021
.\"    Manual: LinuxCNC Documentation
.\"    Source: LinuxCNC
.\"  Language: English
.\"
.TH "SENDKEYS" "1" "04/30/2021" "LinuxCNC" "LinuxCNC Documentation"
.\" -----------------------------------------------------------------
.\" * Define some portability stuff
.\" -----------------------------------------------------------------
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.\" http://bugs.debian.org/507673
.\" http://lists.gnu.org/archive/html/groff/2009-02/msg00013.html
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\" -----------------------------------------------------------------
.\" * set default formatting
.\" -----------------------------------------------------------------
.\" disable hyphenation
.nh
.\" disable justification (adjust text to left margin only)
.ad l
.\" -----------------------------------------------------------------
.\" * MAIN CONTENT STARTS HERE *
.\" -----------------------------------------------------------------
.SH "NAME"
sendkeys \- send input events based on pins or scancodes from HAL
.SH "SYNOPSIS"
.sp
\fBloadusr sendkeys config=s8t5, 16, t12\fR
.SH "DESCRIPTION"
.sp
This component is intended as a partner component to matrix_kb or the hostmot2 7i73 driver\&. It accepts the key\-up and key\-down event codes from either of these and converts them to keystokes sent from a virtual keyboard\&.
.sp
It also allows for keystrokes to be generated by individual HAL pins\&.
.sp
The \fBconfig\fR parameter to the \fBloadusr\fR HAL command defines how many scancodes will be supported and how many individual pins are created\&. \fBconfig=s16\fR would support the 16 scancodes of a 4x4 matrix\&. \fBconfig=t10\fR would create 10 individual HAL pin triggers\&. \fBconfig=s16t10\fR would create one instance with both the above\&.
.sp
Multiple configs separated by commas will create multiple instances of the component\&. The accepted codes can be seen in the extract from the linux headers here: http://wiki\&.linuxcnc\&.org/cgi\-bin/wiki\&.pl?Scancodes
.sp
The component requires the user to have write permissions to /dev/uinput which is not available by default\&. To give access:
.sp
Create the uinput group and add the LinuxCNC user to it:
.sp
.if n \{\
.RS 4
.\}
.nf
sudo groupdadd \-f uinput
sudo gpasswd \-a username uinput
.fi
.if n \{\
.RE
.\}
.sp
Create a new entry in \&.B/etc/udev/rules\&.d/99\-input\&.rules
.sp
.if n \{\
.RS 4
.\}
.nf
sudo echo KERNEL=="uinput", GROUP="uinput", MODE:="0660" | sudo tee /etc/udev/rules\&.d/88\-input\&.rules
.fi
.if n \{\
.RE
.\}
.sp
Then reboot the machine\&. You can test that is has worked:
.sp
.if n \{\
.RS 4
.\}
.nf
ls \-l /dev/uinput
crw\-rw\-\-\-\- 1 root uinput 10, 223 Nov 11 15:35 /dev/uinput
.fi
.if n \{\
.RE
.\}
.sp
It is possible to link the 7i73 codes to both the matrix_kb comp and this comp, so that some codes operate HAL pins and some send keystrokes\&. Where the option exists it is \fIMUCH\fR better to use HAL pins for things like jogging and machine control\&. This component should really be used only for text entry and GUI operations\&.
.sp
Each key on the matrix is allocated a scan code\&. The simplest way to configure the component is to load the component and open a halmeter showing sendkeys\&.0\&.current\-event\&. Note the code for each physical key\&. (If keys do not give consistent results then you probably need to toggle the value of the matrix_kb\&.0\&.negative\-logic pin and/or invert io pins)\&.
.sp
Then edit the HAL file to assign a key event to each scancode\&. For example:
.sp
setp sendkeys\&.0\&.scan\-event\-21 34
.sp
To set a button to type the letter "G" The key events related to each physical key need to be set up prior to the component activating, but after the component is loaded\&.
.sp
To achieve this there is a pin \fBsendkeys\fR\&./N/\&.\fBinit\fR which should be set to "true" once the events to be sent for each scancode and pin have been set up\&.
.sp
To generate keystrokes from other sources note that a keydown is simply 0xC0 & keycode and keyup is 0x80 & keycode\&.
.SH "PINS"
.sp
\fBsendkeys\&.N\&.keycode\fR u32 in Connect to scancode generator\&.
.sp
\fBsendkeys\&.N\&.current\-event\fR s32 out shows the current scancode without keyup / keydown markers\&.
.sp
\fBsendkeys\&.N\&.init\fR bit in set this pin TRUE once all the event parameters have been set\&.
.SH "PARAMETERS"
.sp
\fBsendkeys\&.N\&.scan\-event\-MM\fR u32 in assign the uinput event codes associated with each scancode\&.
.sp
\fBsendkeys\&.N\&.pin\-event\-MM\fR u32 in assign the uinput codes associated with each HAL bit pin\&.
.SH "EXAMPLE"
.sp
.if n \{\
.RS 4
.\}
.nf
loadusr \-W sendkeys config=16t2
net scancodes hm2_7i73\&.0\&.0\&.keycode => sendkeys\&.0\&.keycode

setp sendkeys\&.0\&.scan\-event\-00 34 # Key G
setp sendkeys\&.0\&.scan\-event\-01  2 # Key 1
setp sendkeys\&.0\&.scan\-event\-02  3 # Key 2
setp sendkeys\&.0\&.scan\-event\-03  4 # Key 3
setp sendkeys\&.0\&.scan\-event\-04 50 # Key M
setp sendkeys\&.0\&.scan\-event\-05 05 # Key 4
setp sendkeys\&.0\&.scan\-event\-06 06 # Key 5
setp sendkeys\&.0\&.scan\-event\-07 07 # Key 6
setp sendkeys\&.0\&.scan\-event\-08 31 # Key S
setp sendkeys\&.0\&.scan\-event\-09  8 # Key 7
setp sendkeys\&.0\&.scan\-event\-10  9 # Key 8
setp sendkeys\&.0\&.scan\-event\-11 10 # Key 9
setp sendkeys\&.0\&.scan\-event\-12 20 # Key T
setp sendkeys\&.0\&.scan\-event\-13 11 # Key 0
setp sendkeys\&.0\&.scan\-event\-14 52 # Key Dot
setp sendkeys\&.0\&.scan\-event\-15 14 # Backspace
setp sendkeys\&.0\&.pin\-event\-00 29 # Left Ctrl
setp sendkeys\&.0\&.pin\-event\-01 57 # Space
setp sendkeys\&.0\&.init 1

#Send Ctl + Space from one trigger
net clear\-errors parport\&.0\&.pin\&.00\&.in sendkeys\&.0\&.trigger\-00 sendkeys\&.0\&.trigger\-01
.fi
.if n \{\
.RE
.\}
.SH "AUTHOR"
.sp
andypugh
.SH "LICENSE"
.sp
GPL\-2\&.0+


docs/NEWS
check debian/changelog for recent news

# some of EMC2 History & News
#############################

2009.06.13 - mshaver added emc2/configs/smithy directory and related files
             for Smithy Company's line of cnc machines.

2006.11.17 - removed HAL types s8, u8, s16, and u16.  HAL now uses float
	     for analog values, bit for booleans, and s32 or u32 for
	     integer values.  The only users who will be affected are
	     those with Universal Stepper Controller cards.  See item 9 at
	     http://wiki.linuxcnc.org/cgi-bin/emcinfo.pl?UpdatingConfigurationsForDevelopmentVersions

2006.11.15 - merged classicladder 0.7.100

2006.11.04 - added a feed-hold HAL pin.  Added M codes to enable/disable
	     feed override, spindle override, feed hold, and adaptive feed
	     individually.  Feed override is M50, spindle override is M51,
	     adaptive feed is M52, and feed hold is M53.  G50/51 are no
	     longer used to enable/disable adaptive feed, use M52 instead.

2006.10.24 - added support for block delete as defined by Section 2.2.2
             in the RS274NGC_3 specs.

2006.10.24 - added support for optional program stop as defined by Section 
             2.2.3 in the RS274NGC_3 specs.

2006.10.14 - manpages have been created for most rtapi and hal calls

2006.10.01 - added a userspace simulator which includes rtapi threads.
	     Use "--enable-simulator" to enable this mode, which cannot
	     be used to control real hardware

2006.09.27 - added support to optionally save joint positions between
             runs ([TRAJ] POSITION_FILE = position.txt)

2006.09.26 - added support to specify UNITS in literal in the ini. 
             (e.g. [TRAJ] LINEAR_UNITS = mm )

2006.09.09 - fixed bug #1384883, optional stop was not optional.

2006.08.09 - AXIS is now a part of EMC2, not a separate source download.

2006.08.07 - Added a spindle-speed override. This allows to override the spindle
	     speed during a program run. The MIN_SPINDLE_OVERRIDE and 
	     MAX_SPINDLE_OVERRIDE in the ini assure that speed doesn't get 
	     reduced or increased too much. Both values are in %.

2006.08.06 - Documentation can be built from source with the 'make docs' target

2006.06.11 - added realtime support for jogwheels.  Three HAL pins per axis:
	     one accepts raw counts from the wheel, one sets the distance moved
	     per count, and one enables that axis to move.

2006.05.17 - added 'adaptive feed', a HAL pin that can be used to scale the
	     feedrate in real time.  Simular to feedrate override, but can
	     only vary from 0.0 to 1.0, and reacts much faster.  Intended 
	     for use with EDM and other applications that require closed 
	     loop control of feedrate.

2006.03.18 - A new pin "count" exists for freqgen.  It is the same as the
             internal symbol "count" shown in the hal introduction.pdf,
             or the result of up OR down if there was an "OR" component.

2006.03.16 - Modified the motion control module to behave more like other
             HAL modules - it no longer adds its functions to threads
             automatically.  All the sample configs have been changed to
             explicitly add the functions, users who have custom configs
             will need to do the same.

2006.03.12 - If a copy of axis 1.3a0 or newer is detected inside src/
             it will be automatically built and installed for
             --enable-run-in-place use

2006.03.10 - minor bugfix in tkemc for jogging

2006.03.10 - added genhexkins, with a minitetra sample config.
             This proves EMC2 works just as good with hexapods as EMC1 did.

2006.03.10 - added VTI support & sample configs

2006.03.07 - lots of fixes to the TP (most were wrong in the old one as well)

2006.03.05 - new Trajectory Planner

2006.02.20 - added a manpage for the runscript.

2006.02.12 - Fixed a bug that was causing the RT code to run very slowly,
             forcing users to set BASE_PERIOD very long.

2006.02.10 - The build process has been replaced, no more recursive $(MAKE)

2006.01.28 - fixed bug #1205237, mode switching sometimes hung

2006.01.08 - M62/M63 digital outputs synchronous to motion
             M64/M65 digital outputs that get set right away

2006.01.06 - work on halconfig.tcl, it is used to display/configure HAL data
             requires the BWidget package, realtime, and HAL to run.

2006.01.04 - fix bug #1387740, a motion that goes nowhere clears feedrate

2006.01.01 - the config picker has i18n support, a German translation exists

2005.12.20 - a new config picker has been written, if you start just the
             runscript this one helps you choose a config to run

2005.12.13 - new runscript, and config organization
             work towards the first release of EMC2

2005.xx.xx - EMC2 now supports tool-changing in complex scenarios

2005.xx.xx - classicladder now a HAL component, a software PLC

2004.xx.xx - HAL (the Hardware Abstraction Layer) is used for
	     addressing hardware drivers, and allows for simple configuration


#developers: this file should hold a history of improvements, bugfixes, etc
             it is useful for keeping track of development


src/hal/user_comps/sendkeys.c
//
//   Copyright (C) 2021 Andy Pugh
//
//   This program is free software; you can redistribute it and/or modify
//   it under the terms of the GNU General Public License as published by
//   the Free Software Foundation; either version 2 of the License, or
//   (at your option) any later version.
//
//   This program is distributed in the hope that it will be useful,
//   but WITHOUT ANY WARRANTY; without even the implied warranty of
//   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//   GNU General Public License for more details.
//
//   You should have received a copy of the GNU General Public License
//   along with this program; if not, write to the Free Software
//   Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301 USA
//

#include "rtapi.h"
#include "hal.h"
#include <linux/uinput.h>
#include <unistd.h>
#include <fcntl.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#define MAX_NAME 12

static int comp_id;

typedef struct{
    hal_u32_t *keycode;
    hal_s32_t *current_event;
    hal_bit_t *init;
    hal_bit_t **trigger;
    hal_u32_t *event;
} sendkeys_hal;

typedef struct {
    int num_codes;
    int num_events;
    int num_triggers;
    bool inited;
    int fd;
    bool *prev;
    hal_u32_t oldcode;
} sendkeys_param;

typedef struct {
    sendkeys_hal *hal;
    sendkeys_param *param;
    int num_insts;
} sendkeys ;

static sendkeys *me;

void emit(int fd, int type, int code, int val)
{
   struct input_event ie;

   ie.type = type;
   ie.code = code;
   ie.value = val;
   write(fd, &ie, sizeof(ie));
}

static void exit_handler(int sig) {
    printf("sendkeys: exiting\n");
    exit(0);
}

static void call_hal_exit(void) {
    hal_exit(comp_id);
}

int init(int argc, char* argv[]){
    int *codes;
    int *pins;
    int i, j;
    comp_id = hal_init("sendkeys");
    if (comp_id < 0) {
        rtapi_print_msg(RTAPI_MSG_ERR, "sendkeys: ERROR: hal_init() failed\n");
        return -1;
    }

    me = hal_malloc(sizeof(sendkeys));
    codes = malloc(sizeof(int));
    pins = malloc(sizeof(int));
    // This is all because RTAPI_MP_ARRAY_** macros do not appear to work in usersapce
    for (i = 1; i < argc; i++){
        static int index = -1;
        static int type = -1;
        int ptr = 0;
        bool shift = 0;
        if (strncmp(argv[i], "config=", 7) == 0) type = 1, ptr = 6;
        if (strncmp(argv[i], "names=", 6) == 0) type = 2, ptr = 5;
        switch (type) {
        case 1: // config
            while (argv[i][ptr]){
                switch (argv[i][ptr]){
                case '=':
                case ' ':
                case ',':
                    index++;
                    codes = realloc(codes, (index + 1) * sizeof(int));
                    pins = realloc(pins, (index + 1) * sizeof(int));
                    codes[index] = 0;
                    pins[index] = 0;
                    shift = 0;
                    break;
                case 's':
                case 'S':
                    shift = 0;
                    break;
                case 't':
                case 'T':
                    shift = 1;
                    break;
                case '0':
                case '1':
                case '2':
                case '3':
                case '4':
                case '5':
                case '6':
                case '7':
                case '8':
                case '9':
                    if (shift){
                    pins[index] = pins[index] * 10 + argv[i][ptr] - '0';
                    } else {
                    codes[index] = codes[index] * 10 + argv[i][ptr] - '0';
                    }
                    break;
                }
                ptr++;;
            }
            me->num_insts = index + 1;
            break;
        case 2: //  Parsing of "names" would be here, but is not enabled yet
            while (argv[i][ptr]){
                ptr++;
            }
            break;
        }
    }
    me->hal = (sendkeys_hal*)hal_malloc(me->num_insts * sizeof(sendkeys_hal));
    me->param = malloc(me->num_insts * sizeof(sendkeys_param));
    for (i = 0; i < me->num_insts; i++){
        sendkeys_hal* hal = &(me->hal[i]);
        sendkeys_param* param = &(me->param[i]);
        if (hal_pin_u32_newf(HAL_IN, &(hal->keycode), comp_id,
        "sendkeys.%i.keycode", i) < 0) {
        rtapi_print_msg(RTAPI_MSG_ERR, "sendkeys.N.keycode error\n");
        return -ENOMEM;}
        if (hal_pin_s32_newf(HAL_OUT, &(hal->current_event), comp_id,
        "sendkeys.%i.current-event", i) < 0) {
        rtapi_print_msg(RTAPI_MSG_ERR, "sendkeys.N.current-event error\n");
        return -ENOMEM;}
        if (hal_pin_bit_newf(HAL_IN, &(hal->init), comp_id,
        "sendkeys.%i.init", i) < 0) {
        rtapi_print_msg(RTAPI_MSG_ERR, "sendkeys.N.init error\n");
        return -ENOMEM;}
        // event params
        rtapi_print_msg(RTAPI_MSG_DBG, "instance %i events %i triggers %i\n", i, codes[i], pins[i]);
        param->num_triggers = pins[i];
        param->num_codes = codes[i];
        param->num_events = codes[i] + pins[i];
        hal->event = hal_malloc(param->num_events  * sizeof(hal_u32_t*));
        for (j = 0; j < param->num_codes; j++){
            if (hal_param_u32_newf(HAL_RW, &(hal->event[j]), comp_id,
                    "sendkeys.%i.scan-event-%02i", i, j) < 0) {
                rtapi_print_msg(RTAPI_MSG_ERR, "sendkeys.N.scan-event-%02i error\n", j);
                return -ENOMEM;}
        }
        for (j = 0; j < param->num_triggers; j++){
            if (hal_param_u32_newf(HAL_RW, &(hal->event[j + param->num_codes]), comp_id,
                    "sendkeys.%i.pin-event-%02i", i, j) < 0) {
                rtapi_print_msg(RTAPI_MSG_ERR, "sendkeys.N.pin-event-%02i error\n", j);
                return -ENOMEM;}
        }
        // trigger pins
        hal->trigger = hal_malloc(param->num_triggers * sizeof(hal_bit_t*));
        param->prev = malloc(param->num_triggers * sizeof(hal_bit_t));
        for (j = 0; j < param->num_triggers; j++){
            if (hal_pin_bit_newf(HAL_IN, &(hal->trigger[j]), comp_id,
                    "sendkeys.%i.trigger-%02i", i, j) < 0) {
                rtapi_print_msg(RTAPI_MSG_ERR, "sendkeys.N.trigger-%02i error\n",j);
                return -ENOMEM;}
            param->prev[j] = 0;
        }
    }
    return 0;
}

int main(int argc, char* argv[]) {
    struct uinput_user_dev uidev;
    int i, j;
    
    signal(SIGINT, exit_handler);
    signal(SIGTERM, exit_handler);
    atexit(call_hal_exit);

    //parse configs and create pins
    init(argc, argv);
    hal_ready(comp_id);
    
    while (1) {
        for (i = 0; i < me->num_insts; i++){
            sendkeys_hal* hal = &(me->hal[i]);
            sendkeys_param* param = &(me->param[i]);
            if ((*hal->keycode & 0xC0) == 0x80){
                *hal->current_event = (*hal->keycode & 0x3f);
            } else {
                *hal->current_event = -1;
            }
            // The event codes are written to pins in HAL after the 
            // component is loaded.
            // We need to set up the keyboard based on the events selected
            // so there is a secondary init here.
            if (! *hal->init) continue;
            if (*hal->init && ! param->inited) {
                param->fd = open("/dev/uinput", O_WRONLY | O_NONBLOCK);
                if (param->fd < 0) rtapi_print_msg(RTAPI_MSG_ERR, 
                    "Cannot open /dev/uinput. Suggest chmod 666 /dev/uinput\n");
                ioctl(param->fd, UI_SET_EVBIT, EV_KEY);
                for (j = 0; j < param->num_events; j++){
                    if (hal->event[j] > 0 && hal->event[j] < KEY_MAX){
                        ioctl(param->fd, UI_SET_KEYBIT, hal->event[j]);
                        rtapi_print("SET_EVBIT %i\n", hal->event[j]);
                    }
                }
                memset(&uidev, 0, sizeof(uidev));
                strcpy(uidev.name, "linuxcnc-hal");
                uidev.id.bustype = BUS_USB;
                uidev.id.vendor  = 0x1;
                uidev.id.product = 0x1;
                uidev.id.version = 1;

                write(param->fd, &uidev, sizeof(uidev));
                ioctl(param->fd, UI_DEV_CREATE);
                param->inited = 1;
            }
            if (*hal->keycode != param->oldcode) {
                /* Key press, report the event, send key release, and report again*/
                if ((*hal->keycode & 0x3F) > param->num_events) continue;
                if (hal->event[*hal->keycode & 0x3F] == 0) continue;
                if ((*hal->keycode & 0xC0) == 0xC0){ // keydown
                    emit(param->fd, EV_KEY, hal->event[*hal->keycode & 0x3F], 1);
                    emit(param->fd, EV_SYN, SYN_REPORT, 0);
                } else if ((*hal->keycode & 0xC0) == 0x80){ // keyup
                    emit(param->fd, EV_KEY, hal->event[*hal->keycode & 0x3F], 0);
                    emit(param->fd, EV_SYN, SYN_REPORT, 0);
                }
                param->oldcode = *hal->keycode;
            }
            for (j = 0; j < param->num_triggers; j++){
                if (param->prev[j] != *hal->trigger[j]){
                    if (*hal->trigger[j]){ // keydown
                        emit(param->fd, EV_KEY, hal->event[param->num_codes + j], 1);
                        emit(param->fd, EV_SYN, SYN_REPORT, 0);
                    } else { // keyup
                        emit(param->fd, EV_KEY, hal->event[param->num_codes + j], 0);
                        emit(param->fd, EV_SYN, SYN_REPORT, 0);
                    }
                    param->prev[j] = *hal->trigger[j];
                }
            }
        }
    }
}
